### ISSUE TITLE
Support SSH hostkey rotation via OpenSSH extensions

### ISSUE THREAD

**Issue by Castaglia** [proftpd/proftpd#1323]
The scope of this issue is to implement support for [graceful SSH hostkey rotation](https://lwn.net/Articles/637156/) via the following OpenSSH extensions (see [here](https://github.com/openssh/openssh-portable/blob/master/PROTOCOL#L286)):
* `hostkeys-00@openssh.com`
* `hostkeys-prove-00@openssh.com`

This support will be configurable via `SFTPOption`.

**Comment by Castaglia** [comment]
Implemented in master.

**Comment by severach** [comment]
It's been a challenge to figure out how any of this works. The check happens after authenticated logon. The server has two RSA keys 1024 and 3072.

Is this a bug?

```
# ssh -V
OpenSSH_9.0p1, OpenSSL 1.1.1o  3 May 2022
#sftp -P 2222 -v chris@127.0.0.1
chris@127.0.0.1's password: 
Authenticated to 127.0.0.1 ([127.0.0.1]:2222) using "password".
debug1: channel 0: new [client-session]
debug1: Entering interactive session.
debug1: pledge: filesystem
debug1: client_input_global_request: rtype hostkeys-00@openssh.com want_reply 0
debug1: client_input_hostkeys: convert key: unknown or unsupported key type
debug1: client_input_hostkeys: searching /root/.ssh/known_hosts for [127.0.0.1]:2222 / (none)
debug1: client_input_hostkeys: searching /root/.ssh/known_hosts2 for [127.0.0.1]:2222 / (none)
debug1: client_input_hostkeys: hostkeys file /root/.ssh/known_hosts2 does not exist
debug1: client_input_hostkeys: host key found matching a different name/address, skipping UserKnownHostsFile update
debug1: Sending subsystem: sftp
```

**Comment by norrisjeremy** [comment]
I don't believe you can configure a server with multiples of the same key type (i.e., you can't configure a server with two different RSA keys).

**Comment by norrisjeremy** [comment]
> I don't believe you can configure a server with multiples of the same key type (i.e., you can't configure a server with two different RSA keys).

To expand on what I mean:

Server is turned up with an RSA key & DSA key, which are both populated into client's known_hosts file.

Later the server administrator decides to reconfigure server to no longer have a DSA key, and instead have an Ed25519 key (whilst still having the same original RSA key).

When a client connects after the server administrator has made this change, it authenticates server using the original RSA key that is present in the known_hosts file, then via the hostkey rotation extension, removes the old DSA key from the known_host file and adds the new Ed25519 key.

I don't believe the hostkey rotation extension solves for the case to allow for rotation of the same underlying key type (i.e., having two different RSA keys, two different Ed25519 keys, etc.).

**Comment by severach** [comment]
>OpenSSH 6.8: Sometimes, the key replacement is precautionary (such as migrating to a stronger key algorithm), but key replacement may also be necessary in a hurry if a key is believed to be compromised.

That may be how it's implemented but that's not the intent. They can't be suggesting that I'd switch to Ed25519 if my RSA key was compromised. Instead I'd be switching to another RSA key, maybe larger, using the old one just long enough to get my clients updated. What do I do if my Ed25519 key is compromised? What if they are all compromised?

>OpenSSH 6.8: The server, in turn, can eventually pull an old key out of the list and retire it. The client, when it connects with the new key, would update its list again and remove the now-absent old key from known_hosts.

For this to work as you suggest, I'd need move all the clients to my new Ed25519 key, update the RSA key, then move them back to RSA.

[derobert](https://unix.stackexchange.com/questions/334597/how-to-roll-over-ssh-host-keys) wants what I want.

**Comment by norrisjeremy** [comment]
Yes, it appears you are correct, I was misremembering that they do appear to at least document support for this use case according to [section 2.5](https://github.com/openssh/openssh-portable/blob/master/PROTOCOL#L333) of their PROTOCOL document.

```
These extensions let a client learn key types that it had not previously
encountered, thereby allowing it to potentially upgrade from weaker
key algorithms to better ones. It also supports graceful key rotation:
a server may offer multiple keys of the same type for a period (to
give clients an opportunity to learn them using this extension) before
removing the deprecated key from those offered.
```

Whether that's true in practice or not, I'm not certain (I've not tested or read through the code enough to confirm).

**Comment by Castaglia** [comment]
@severach This particular message from your client is what I'm most interested in at the moment:
```
debug1: client_input_hostkeys: convert key: unknown or unsupported key type
```
Which particular types of hostkeys (and sizes) are configured in your `mod_sftp`, when the above error is logged by the client?
